py/makemoduledefs.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+
+# This pre-processor parses provided objects' c files for 
+# `MP_REGISTER_MODULE(module_name, obj_module, enabled_define)`
+# These are used to generate a header with the required entries for 
+# `mp_rom_map_elem_t mp_builtin_module_table[]` in py/objmodule.c
+
+from __future__ import print_function
+
+import re
+import os
+import argparse
+
+
+def find_c_file(vpath, obj_file):
+    """
+    Search the provided vpaths for the c file that matches the provided object_file
+    :param List[str} vpath: List of base paths, similar to gcc vpath
+    :param obj_file: object file to find the matching c file for
+    :return: path to c file or None
+    """
+    c_file = None
+    relative_c_file = os.path.splitext(obj_file)[0] + ".c"
+    relative_c_file = relative_c_file.lstrip('/\\')
+    for p in vpath:
+        possible_c_file = os.path.join(p, relative_c_file)
+        if os.path.exists(possible_c_file):
+            c_file = possible_c_file
+            break
+
+    return c_file
+
+
+def find_module_registrations(c_file):
+    """
+    Find any MP_REGISTER_MODULE definitions in the provided c file
+    :param c_file: path to c file to check
+    :return: List((module_name, obj_module, enabled_define))
+    """
+    pattern = re.compile(
```

Comment:
Should be closer to where it's used, or maybe don't use seperate pattern at all

Reply:
Ah in previous structure this pattern compilation happened outside the for loop iterating over files, so should be a decent efficiency improvement. It's now essentially moved inside the loop though so the benefit has been lost.

I'm not sure just how much is gained by only compiling this once up front, rather than once per file... so not too sure if it's worth moving it to global scope to keep it separate.

Reply:
Sorry should have been clearer, I meant: perhaps it's clearer and more concies to simply not compile to a pattern at all, but just pass the regex string as first argument to re.findall.

Reply:
> perhaps it's clearer and more concies to simply not compile to a pattern at all, but just pass the regex string as first argument to re.findall.

Well, nope. Compiling patterns is the best practice, shortcuts like that exist only for once-off scripts.

Reply:
Yeah I prefer the once off compilation, my last push moved the pattern compilation to global scope to ensure it's a once off.

Reply:
>  How much speed though, I'm not sure.

Yes that's what struck me first time. De-clarifying code just for the sake of 'might be faster' in some specific Python version in a program which is likely going to be limited by disk I/O, don't know. Well, it won't hurt either.